engine/table/src/main/java/io/deephaven/engine/table/impl/sources/UnionSourceManager.java

```diff
@@ -499,10 +499,12 @@ private void processExisting(@NotNull final Table constituent) {
             }
 
             if (changes == null || changes.empty()) {
+                // constituent is either static or did not change this cycle
                 if (slotAllocationChanged) {
                     currFirstRowKeys[nextCurrentSlot + 1] = checkOverflow(nextSlotPrevFirstRowKey + shiftDelta);
                     resultRows.insertWithShift(currFirstRowKey, constituent.getRowSet());
-                    if (shiftDelta != 0) {
+                    // don't bother shifting if the constituent is empty
+                    if (shiftDelta != 0 && constituent.size() > 0) {
```

Comment:
We should move the line above (`insertWithShift`) into this `if` block.

Reply:
Is `shiftDelta` guaranteed to be non-zero here? We clear `resultRows` past the first shifted row when we set `slotAllocationChanged` to true. If it is valid for shiftDelta to be zero, then we must add the constituent to the result set unconditionally.

I'll drop the `shiftDelta` of zero check. I believe with your partition table related changes that constituents can be removed causing negative shifts; and thus a shift of zero is possible. However, `shiftRange` does ignore shifts of zero.